substrate/executor/src/native_executor.rs

```diff
@@ -17,10 +17,32 @@
 use error::{Error, ErrorKind, Result};
 use state_machine::{CodeExecutor, Externalities};
 use wasm_executor::WasmExecutor;
+use wasmi::Module as WasmModule;
 use runtime_version::RuntimeVersion;
+use std::collections::HashMap;
 use codec::Slicable;
+use twox_hash::XxHash;
+use std::hash::Hasher;
+use parking_lot::Mutex;
 use RuntimeInfo;
 
+// For the internal Runtime Cache
+enum RunWith {
+	NativeRuntime,
+	WasmRuntime(WasmModule)
+}
+
+lazy_static! {
+	static ref RUNTIMES_CACHE: Mutex<HashMap<u64, RunWith>> = Mutex::new(HashMap::new());
+}
+
+// helper function to generate low-over-head caching_keys
+fn gen_cache_key(code: &[u8]) -> u64 {
+	let mut h = XxHash::with_seed(0);
```

Comment:
since `XxHash` is not crypto-secure, it's worth putting a comment here about why we don't consider it possible that two different code preimages will be able to hash to the same `u64`. Something like "it is asserted that part of the audit process that any potential on-chain code change will have done is to ensure that the two-x hash is different to that of any other `:code` value from the same chain".